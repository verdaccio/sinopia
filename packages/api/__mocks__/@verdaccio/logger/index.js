const debug = require('debug')('verdaccio:test');

const setup = debug;
const logger = {
  child: jest.fn(() => ({
    debug,
    trace: debug,
    warn: debug,
    info: debug,
    error: debug,
    fatal: debug,
  })),
  debug: debug,
  trace: debug,
  warn: debug,
  info: debug,
  error: debug,
  fatal: debug,
};

export { setup, logger };
