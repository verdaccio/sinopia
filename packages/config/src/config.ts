import assert from 'assert';
import _ from 'lodash';
import buildDebug from 'debug';

import {
  getMatchedPackagesSpec,
  normalisePackageAccess,
  sanityCheckUplinksProps,
  uplinkSanityCheck,
  generateRandomHexString,
  getUserAgent,
  isObject,
} from '@verdaccio/utils';
import { APP_ERROR } from '@verdaccio/dev-commons';

import {
  PackageList,
  Config as AppConfig,
  Security,
  Logger,
  PackageAccess,
} from '@verdaccio/types';
import { generateRandomSecretKey } from './token';

const LoggerApi = require('@verdaccio/logger');

const strategicConfigProps = ['uplinks', 'packages'];
const allowedEnvConfig = ['http_proxy', 'https_proxy', 'no_proxy'];

export type MatchedPackage = PackageAccess | void;

export interface StartUpConfig {
  storage: string;
  plugins?: string;
  self_path: string;
}

const debug = buildDebug('verdaccio:config');

/**
 * Coordinates the application configuration
 */
class Config implements AppConfig {
  public logger: Logger;
  public user_agent: string;
  // @ts-ignore
  public secret: string;
  public uplinks: any;
  public packages: PackageList;
  public users: any;
  public server_id: string;
  public self_path: string;
  public storage: string | void;
  public plugins: string | void;
  // @ts-ignore
  public security: Security;

  public constructor(config: StartUpConfig) {
    const self = this;
    this.logger = LoggerApi.logger;
    this.self_path = config.self_path;
    this.storage = config.storage;
    this.plugins = config.plugins;

    for (const configProp in config) {
      if (self[configProp] == null) {
        self[configProp] = config[configProp];
      }
    }

    // @ts-ignore
    if (_.isNil(this.user_agent)) {
      this.user_agent = getUserAgent();
    }

    // some weird shell scripts are valid yaml files parsed as string
    assert(_.isObject(config), APP_ERROR.CONFIG_NOT_VALID);

    // sanity check for strategic config properties
    strategicConfigProps.forEach(function (x): void {
      if (self[x] == null) {
        self[x] = {};
      }

      assert(isObject(self[x]), `CONFIG: bad "${x}" value (object expected)`);
    });

    this.uplinks = sanityCheckUplinksProps(uplinkSanityCheck(this.uplinks));

    if (_.isNil(this.users) === false) {
      this.logger.warn(`[users]: property on configuration file
      is not longer supported, property being ignored`);
    }

    this.packages = normalisePackageAccess(self.packages);

    // loading these from ENV if aren't in config
    allowedEnvConfig.forEach((envConf): void => {
      if (!(envConf in self)) {
        self[envConf] = process.env[envConf] || process.env[envConf.toUpperCase()];
      }
    });

    // unique identifier of self server (or a cluster), used to avoid loops
    // @ts-ignore
    if (!this.server_id) {
      this.server_id = generateRandomHexString(6);
    }
  }

  /**
   * Check for package spec
   */
  public getMatchedPackagesSpec(pkgName: string): MatchedPackage {
    return getMatchedPackagesSpec(pkgName, this.packages);
  }

  /**
   * Store or create whether receive a secret key
   */
  public checkSecretKey(secret: string): string {
    debug('check secret key');
    if (_.isString(secret) && _.isEmpty(secret) === false) {
      this.secret = secret;
      debug('reusing previous key');
      return secret;
    }
    // it generates a secret key
    // FUTURE: this might be an external secret key, perhaps within config file?
    debug('generate a new key');
    this.secret = generateRandomSecretKey();
    return this.secret;
  }
}

export { Config };
