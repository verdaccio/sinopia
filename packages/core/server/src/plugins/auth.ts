import { Auth, IAuth } from '@verdaccio/auth';

export async function authService(fastify, opts, done) {
  const { config } = opts;
  const auth: IAuth = new Auth(config);
  fastify.decorate('auth', auth);
  done();
}
