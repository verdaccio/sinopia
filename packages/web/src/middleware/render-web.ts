import buildDebug from 'debug';
import express from 'express';
import _ from 'lodash';
import path from 'path';

import { HTTP_STATUS } from '@verdaccio/core';
import { asyncLoadPlugin } from '@verdaccio/loaders';
import { logger } from '@verdaccio/logger';
import { isURLhasValidProtocol } from '@verdaccio/url';

import renderHTML from '../renderHTML';
import { setSecurityWebHeaders } from './security';

const debug = buildDebug('verdaccio:web:render');

export async function loadTheme(config: any) {
  if (_.isNil(config.theme) === false) {
    const plugin = await asyncLoadPlugin(
      config.theme,
      { config, logger },
      function (plugin) {
        return typeof plugin === 'string';
      },
      config?.serverSettings?.pluginPrefix ?? 'verdaccio-theme'
    );
    if (plugin.length > 1) {
      logger.warn(
        'multiple ui themes has been detected and is not supported, only the first one will be used'
      );
    }

    return _.head(plugin);
  }
}

const sendFileCallback = (next) => (err) => {
  if (!err) {
    return;
  }
  if (err.status === HTTP_STATUS.NOT_FOUND) {
    next();
  } else {
    next(err);
  }
};

export async function renderWebMiddleware(config, auth): Promise<any> {
  const { staticPath, manifest, manifestFiles } =
    (await loadTheme(config)) || require('@verdaccio/ui-theme')();
  debug('static path %o', staticPath);

  /* eslint new-cap:off */
  const router = express.Router();
  router.use(auth.webUIJWTmiddleware());
  router.use(setSecurityWebHeaders);

  // Logo
  let logoURI = config?.web?.logo ?? '';
  if (logoURI && !isURLhasValidProtocol(logoURI)) {
    // URI related to a local file

    // Note: `path.join` will break on Windows, because it transforms `/` to `\`
    // Use POSIX version `path.posix.join` instead.
    logoURI = path.posix.join('/-/static/', path.basename(logoURI));
    router.get(logoURI, function (req, res, next) {
      res.sendFile(path.resolve(config.web.logo), sendFileCallback(next));
      debug('render static');
    });
  }

  // Static
  router.get('/-/static/*', function (req, res, next) {
    const filename = req.params[0];
    const file = `${staticPath}/${filename}`;
    debug('render static file %o', file);
    res.sendFile(file, sendFileCallback(next));
  });

  router.get('/-/web/:section/*', function (req, res) {
    renderHTML(config, manifest, manifestFiles, req, res);
    debug('render html section');
  });

  router.get('/', function (req, res) {
    renderHTML(config, manifest, manifestFiles, req, res);
    debug('render root');
  });

  return router;
}
