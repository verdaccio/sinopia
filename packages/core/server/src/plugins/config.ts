import { FastifyInstance } from 'fastify';
import fp from 'fastify-plugin';

import { Config as AppConfig } from '@verdaccio/config';
import { ConfigRuntime, Config as IConfig } from '@verdaccio/types';

export default fp(
  async function (fastify: FastifyInstance, opts: { config: ConfigRuntime }) {
    const { config } = opts;
    const configInstance: IConfig = new AppConfig(Object.assign({}, config));
    fastify.decorate('configInstance', configInstance);
  },
  {
    fastify: '>=3.x',
  }
);

declare module 'fastify' {
  interface FastifyInstance {
    configInstance: IConfig;
  }
}
