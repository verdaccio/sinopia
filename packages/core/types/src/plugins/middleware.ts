import { Config } from '../configuration';
import { IBasicAuth } from './auth';
import { IPlugin } from './commons';

// TODO: convert to generic storage should come from implementation
export interface IPluginMiddleware<T, K> extends IPlugin<T> {
  register_middlewares(app: any, auth: IBasicAuth<T>, storage: K): void;
}
