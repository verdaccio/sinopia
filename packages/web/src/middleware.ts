import express from 'express';
import _ from 'lodash';

import { PLUGIN_CATEGORY } from '@verdaccio/core';
import { asyncLoadPlugin } from '@verdaccio/loaders';
import { logger } from '@verdaccio/logger';
import { webMiddleware } from '@verdaccio/middleware';

import webEndpointsApi from './api';

export async function loadTheme(config: any) {
  if (_.isNil(config.theme) === false) {
    const plugin = await asyncLoadPlugin(
      config.theme,
      { config, logger },
      // TODO: add types { staticPath: string; manifest: unknown; manifestFiles: unknown }
      function (plugin: any) {
        /**
         * 
          - `staticPath`: is the same data returned in Verdaccio 5.
          - `manifest`: A webpack manifest object.
          - `manifestFiles`: A object with one property `js` and the array (order matters) of the manifest id to be loaded in the template dynamically.
         */
        return plugin.staticPath && plugin.manifest && plugin.manifestFiles;
      },
      config?.serverSettings?.pluginPrefix ?? 'verdaccio-theme',
      PLUGIN_CATEGORY.THEME
    );
    if (plugin.length > 1) {
      logger.warn('multiple ui themes are not supported; only the first plugin is used');
    }

    return _.head(plugin);
  }
}

export default async (config, auth, storage, logger) => {
  let pluginOptions = await loadTheme(config);
  if (!pluginOptions) {
    pluginOptions = require('@verdaccio/ui-theme')(config.web);
    logger.info(
      { name: '@verdaccio/ui-theme', pluginCategory: PLUGIN_CATEGORY.THEME },
      'plugin @{name} successfully loaded (@{pluginCategory})'
    );
  }

  // eslint-disable-next-line new-cap
  const router = express.Router();
  // load application
  router.use(
    webMiddleware(
      config,
      {
        tokenMiddleware: auth.webUIJWTmiddleware(),
        webEndpointsApi: webEndpointsApi(auth, storage, config),
      },
      pluginOptions
    )
  );
  return router;
};
