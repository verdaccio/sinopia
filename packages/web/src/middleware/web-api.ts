import bodyParser from 'body-parser';
import { Router } from 'express';

import { Auth } from '@verdaccio/auth';
import { validateName, validatePackage } from '@verdaccio/middleware';
import { Storage } from '@verdaccio/store';
import { Config } from '@verdaccio/types';

import webEndpointsApi from '../api';
import { setSecurityWebHeaders } from './security';

export function webAPI(config: Config, auth: Auth, storage: Storage): Router {
  // eslint-disable-next-line new-cap
  const route = Router();
  // validate all of these params as a package name
  // this might be too harsh, so ask if it causes trouble=
  route.param('package', validatePackage);
  route.param('filename', validateName);
  route.param('version', validateName);
  route.use(bodyParser.urlencoded({ extended: false }));
  route.use(auth.webUIJWTmiddleware());
  route.use(setSecurityWebHeaders);
  route.use(webEndpointsApi(auth, storage, config));
  return route;
}
