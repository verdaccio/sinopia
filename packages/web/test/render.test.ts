import supertest from 'supertest';
import express from 'express';
import path from 'path';
import { Application } from 'express';
import bodyParser from 'body-parser';
import { Config, parseConfigFile } from '@verdaccio/config';
import { final, handleError, errorReportingMiddleware } from '@verdaccio/middleware';

import { Storage } from '@verdaccio/store';
import { Auth, IAuth } from '@verdaccio/auth';
import { setup } from '@verdaccio/logger';
import { HEADERS, HEADER_TYPE, HTTP_STATUS, API_ERROR } from '@verdaccio/commons-api';
import routes from '../src';
import { parseHtml } from './partials/htmlParser';
import { initializeServer } from './helper';

setup([]);

const mockManifest = jest.fn();
jest.mock('@verdaccio/ui-theme', () => mockManifest());

const getConf = (conf) => {
  const configPath = path.join(__dirname, 'config', conf);

  return parseConfigFile(configPath);
};

describe('test web server', () => {
  beforeAll(() => {
    mockManifest.mockReturnValue({
      staticPath: path.join(__dirname, 'static'),
      manifest: require('./partials/manifest/manifest.json'),
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
    mockManifest.mockClear();
  });

  describe('render', () => {
    test('should return the root', async () => {
      return (
        supertest(await initializeServer('default-test.yaml'))
          .get('/')
          .set('Accept', HEADERS.TEXT_HTML)
          // .expect(HEADER_TYPE.CONTENT_TYPE, HEADERS.TEXT_HTML)
          .expect(HTTP_STATUS.OK)
          .then((response) => {
            // eslint-disable-next-line no-console
            // console.log(response.text);
            // eslint-disable-next-line no-console
            // console.log(parseHtml(response.text).querySelectorAll('body').toString());
          })
      );
    });

    test('should return the body for a package detail page', async () => {
      return (
        supertest(await initializeServer('default-test.yaml'))
          .get('/-/web/section/some-package')
          .set('Accept', HEADERS.TEXT_HTML)
          // .expect(HEADER_TYPE.CONTENT_TYPE, HEADERS.TEXT_HTML)
          .expect(HTTP_STATUS.OK)
          .then((response) => {
            // eslint-disable-next-line no-console
            // console.log(response.text);
            // eslint-disable-next-line no-console
            // console.log(parseHtml(response.text).querySelectorAll('body').toString());
          })
      );
    });

    test('should static file not found', async () => {
      return supertest(await initializeServer('default-test.yaml'))
        .get('/-/static/not-found.js')
        .set('Accept', HEADERS.TEXT_HTML)
        .expect(HTTP_STATUS.NOT_FOUND);
    });
    //
    test('should static file found', async () => {
      return supertest(await initializeServer('default-test.yaml'))
        .get('/-/static/main.js')
        .set('Accept', HEADERS.TEXT_HTML)
        .expect(HTTP_STATUS.OK);
    });
  });
});
