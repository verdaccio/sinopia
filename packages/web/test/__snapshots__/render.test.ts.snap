// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test web server render output should match render 1`] = `
"
    <!DOCTYPE html>
      <html lang=\\"en-us\\"> 
      <head>
        <meta charset=\\"utf-8\\">
        <base href=\\"http://127.0.0.1:51411/\\">
        <title>verdaccio</title>        
        <link rel=\\"icon\\" href=\\"http://127.0.0.1:51411/-/static/favicon.ico\\"/>
        <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" /> 
        <script>
            window.__VERDACCIO_BASENAME_UI_OPTIONS={\\"darkMode\\":false,\\"basename\\":\\"/\\",\\"base\\":\\"http://127.0.0.1:51411/\\",\\"primaryColor\\":\\"#4b5e40\\",\\"version\\":\\"6.0.0-6-next.28\\",\\"logoURI\\":\\"\\",\\"flags\\":{\\"searchRemote\\":true},\\"login\\":true,\\"pkgManagers\\":[\\"yarn\\",\\"pnpm\\",\\"npm\\"],\\"title\\":\\"verdaccio\\",\\"scope\\":\\"\\",\\"language\\":\\"es-US\\"}
        </script>
        
      </head>    
      <body class=\\"body\\">
      
        <div id=\\"root\\"></div>
        <script defer=\\"defer\\" src=\\"http://127.0.0.1:51411//-/static/runtime.6126058572f989c948b1.js\\"></script><script defer=\\"defer\\" src=\\"http://127.0.0.1:51411//-/static/vendors.6126058572f989c948b1.js\\"></script><script defer=\\"defer\\" src=\\"http://127.0.0.1:51411//-/static/main.6126058572f989c948b1.js\\"></script>
        
      </body>
    </html>
  "
`;
